File: javascript/collatz-conjecture/collatz-conjecture.js
//
// This is only a SKELETON file for the 'Collatz Conjecture' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const steps = (number) => {
    let step = 0;
    if (number <= 0) {
        throw new Error('Only positive numbers are allowed');
    }
    for (step; number != 1; step++) {
        number % 2 == 0 ? (number = number / 2) : (number = number * 3 + 1);
    }
    return step;
};

File: dart/collatz-conjecture/lib/collatz_conjecture.dart
class CollatzConjecture {
  int steps(int number) {
    int step = 0;
    onlyPositive(number);
    while (number > 1) {
      checkandupdate(number);
      print(step);
      number = checkandupdate(number);
      step++;
    }
    return step;
  }

  int checkandupdate(int num) => num = num.isEven ? num ~/ 2 : num * 3 + 1;
  void onlyPositive(int num) {
    if (num.isNegative || num == 0) {
      throw ArgumentError('Only positive numbers are allowed');
    }
  }
}
